compiler/rustc_errors/src/lib.rs

```diff
@@ -1289,8 +1299,9 @@ impl DiagCtxtInner {
                     continue;
                 }
             }
-            self.emit_diagnostic(diag);
```

Comment:
This eagerly calls `emit_diagnostic`, looks suspicious.

Reply:
Sorry, I don't understand this. How is it suspicious? What is the alternative to calling it "eagerly"?

Reply:
Sorry, linked wrong line.

`guar = guar.or(self.emit_diagnostic(diag));`

If i see `or` with `Option`, expecting that `or`'s argument will be 1. cheap (for perf reasons, not applied here) 2. without side effects, otherwise using `or` is misleading: you eval arg, but throwing out its result.

Reply:
I see. How would you write it instead?

Reply:
Because the idea here is that every `ErrorGuaranteed` is equivalent. So it doesn't matter which one we use. And it also relies on `or` not short-circuiting.

Reply:
Other problem is that in future someone will see that function call in .or(..) and will refactor into .or_else(..) which will break that `not short-circuiting` expectation here.